without-communication/Projet.ipynb

```diff
@@ -0,0 +1,262 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "a8190f76",
+   "metadata": {},
+   "source": [
+    "# Imports"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "id": "204637ac",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import mesa\n",
+    "from mesa import Agent, Model\n",
+    "from mesa.time import BaseScheduler"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "bc82b2a2",
+   "metadata": {},
+   "source": [
+    "# Agents"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "bd6fc3a1",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\"\"\"\n",
+    "J'imagine percepts comme cela:\n",
+    "percepts = {(x1, y1) : {\"agents\": [list of agents at pos1],\n",
```

Comment:
c'est pas possible car un agent n'a de vision que sur ce qu'il voit. Il ne peut pas connaitre les positions des autres objets sur la map.
De plus, cela voudrait dire que nous avons une variable qui a une taille enorme : dict N_CELL x 3 (agent+waste+zone)  x NB_AGENTS

Reply:
dans le word (et readme), on a définit knowledge comme un dict à deux variables : 
- percepts : tableau des percepts récupéré après chaque action
- action : liste des actions réalisées

ca devrait donc plus ressembler à qqch comme : 
```

knowledges = {
    "actions": [
        Action.STAY,
        Action.RIGHT,
    ],
    "percepts": [
        {  # percept 0
            "radiactivity": 0.11,
            "waste1": "green",
            "waste2": "empty",
            "pos": (1, 2),
            "other_on_pos": True,
        },
        {  # percept 1, after moving RIGHT
            "radiactivity": 0.23,
            "waste1": "green",
            "waste2": "green",
            "pos": (2, 2),
            "other_on_pos": False,
        },
    ],
}
```

Reply:
je vais le préciser dans le readme pour que ca soit plus claire

Reply:
cette structure :
```
 percepts = {pos1 : {contents of case at pos1}, 
             pos2 : {contents of tile at pos2},
                             ...}
```
n'est pas mon idée, ce sont directement les profs qui l'ont proposée. Elle est indiquée dans le sujet, je te laisse regarder. Je n'ai fait que suggérer une manière de remplir les valeurs `{contents of tile at pos i}`. 

Les agents ont accès aux cases environnantes avec la fonction : `grid.get_neighborhood( (x,y), moore=True, include_center=False )`.  Il est également possible d'avoir accès au contenu de ces cases avec `self.model.grid.get_cell_list_contents([self.pos])`. C'est ce qu'on a fait en TP et je pense que c'est fait pour être réutilisé en projet, mais on pourra toujours demander aux profs stv. Et la variable n'est pas énorme, puisqu'un agent  robot n'a accès qu'à ses voisins et pas à toutes les cases de la grille.

Reply:
Pour knowledge, ma variable ressembla à ca : 
`knowledge = {\"color\":\"green\", \"pos\": pos,\"wastes\":[], \"transformed\": False, \"percepts\": {}}`

Si j'ai bien compris ce qui était demandé par le sujet, on doit créer une fonction `deliberate` (et non une méthode par classe de robot) qui prenne en argument l'attribut `knowledge`. Cette fonction devra retourner l'action à prendre par le robot. Cette action dépends fortement du type de robot, d'où l'importance d'avoir la clé `color` dans `knowledge` à mon avis. Ensuite j'ai mis `pos` pour indiquer la position du robot, `wastes` pour indiquer les déchets ramassés par le robot et `transformed` pour indiqué si le déchet est transformé ou pas. 

Pour mettre `actions` dans knowledge, je suis pas sûr de comprendre pourquoi. Si c'est pour tenir un historique des actions par agents, alors c'est le rôle du `DataCollector`. Je veux bien une explication pls.

Reply:
tu n'as pas besoin de l'attribut color car la fonction deliberate doit être implémentée dans chaque classe comme précisée dans le sujet. Donc chaque fonction deliberate implémentée dans un agent se propre à l'agent (qui sait sa couleur).
Pour la variable transformed, je ne suis pas sure qu'elle soit utile puisqu'un robot vert qui a un déchet jaune l'a forcément transformé par exemple, et de même pour un déchet vert qui sera forcément non transformer. Je ne vois pas comment cette variable peut être utilisée afin de délibéré.
Enfin, attention par rapport au `DataCollector`. Cet objet permet de collecter les données du jeu pour les analyser ensuite, mais ni les agents ni l'environnement n'y ont accès. L'idée dans le fait de garder pour chaque robot un historique de ses actions et que, si le robot vient par exemple de monter (action UP), en sachant cela dans sa variable knowledge, au moment de délibéré, il ne va probablement pas descendre tout de suite après. Garder un historique de nos déplacements permettrait de faire une fonction `deliberate` plus intelligente par la suite. C'est pour la même raison qu'il serait intéressant de garder l'historique des percepts dans la knowledge du robot (si on arrive à l'exploiter plus tard pour délibéré)

Reply:
Mais dans un premier temps, fais des agents simple qui avance aléatoirement par exemple, et on les améliorera après.